### PR TITLE
Restrict Auto-Mount of Service Account Tokens in Service Account Object

### LINKED ISSUE #819: [Sample] policies to audit and restrict long-lived tokens for service accounts
### Problem Statement

Prior to v1.24, Kubernetes generated long lived service accounts by default. These pose a security risk.

### Solution Description

Sample policies to:
* audit for secrets with `type: kubernetes.io/service-account-token`
* audit for service accounts with any `secrets.name` configured
* enforce pod `automountServiceAccountToken: false`

### Example "Good" Resource

_No response_

### Example "Bad" Resource

_No response_

### Other Comments

See:
- https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/
- https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/
- https://eng.d2iq.com/blog/service-account-tokens-in-kubernetes-v1.24/

### Slack discussion

https://kubernetes.slack.com/archives/CLGR9BJU9/p1701451535229319

### Troubleshooting

- [X] I have read and followed the documentation AND the [troubleshooting guide](https://kyverno.io/docs/troubleshooting/).
- [X] I have verified the policy does not exist in the [samples library](https://kyverno.io/policies/).

### PR DESCRIPTION
**Pull request by swastik959**
## Related Issue(s)
partially addresses #819 

## Description
added a policy to Restrict Auto-Mount of Service Account Tokens in Service Account Object

## Checklist

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.
- [x] I have added the artifacthub-pkg.yml file and have verified it is complete and correct.

### CODE REVIEW

**Comment by swastik959** [comment]
@chipzoller can you please review this pr

**Comment by chipzoller** [comment]
Please complete PR template.

**Comment by swastik959** [comment]
@chipzoller done

**Review by chipzoller — CHANGES_REQUESTED** [review]
See similar comments from #880

**Comment by swastik959** [comment]
@chipzoller done

**Comment by swastik959** [comment]
@chipzoller please review this pr

File: other/restrict-sa-automount-sa-token/restrict-sa-automount-sa-token.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  - name: validate-sa-automountServiceAccountToken
+    match:
+      any:
+      - resources:
+          kinds:
+          - ServiceAccount
+    validate:
+      message: "Auto-mounting of Service Account tokens is not allowed."
+      pattern:
+        automountServiceAccountToken: "false"
```
Comment (chipzoller) [inline other/restrict-sa-automount-sa-token/restrict-sa-automount-sa-token.yaml:30]: Should be boolean value and not string.

File: other/restrict-sa-automount-sa-token/restrict-sa-automount-sa-token.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  background: true
+  rules:
+  - name: validate-sa-automountServiceAccountToken
+    match:
+      any:
+      - resources:
+          kinds:
+          - ServiceAccount
+    validate:
+      message: "Auto-mounting of Service Account tokens is not allowed."
```
Comment (chipzoller) [inline other/restrict-sa-automount-sa-token/restrict-sa-automount-sa-token.yaml:28]:
```suggestion
      message: "ServiceAccounts must set automountServiceAccountToken to false."
```

File: other/restrict-sa-automount-sa-token/restrict-sa-automount-sa-token.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: restrict-sa-automount-sa-token
+  annotations:
+    policies.kyverno.io/title: Restrict Auto-Mount of Service Account Tokens in Service Account Object
```
Comment (chipzoller) [inline other/restrict-sa-automount-sa-token/restrict-sa-automount-sa-token.yaml:6]:
```suggestion
    policies.kyverno.io/title: Restrict Auto-Mount of Service Account Tokens in Service Accounts
```

**Review by chipzoller — CHANGES_REQUESTED** [review]
Also, `policies.kyverno.io/subject` annotation is missing.

**Comment by swastik959** [comment]
@chipzoller please check it now

**Comment by chipzoller** [comment]
In the future, please simply re-request review after changes have been made. You do not need to keep mentioning me as I subscribe to updates on PRs where changes were requested.

File: other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+createdAt: "2024-01-07T16:15:06.000Z"
+description: >-
+      Kubernetes automatically mounts ServiceAccount credentials in each ServiceAccount.
+      The ServiceAccount may be assigned roles allowing Pods to access API resources.
+      Blocking this ability is an extension of the least privilege best practice and should
+      be followed if Pods do not need to speak to the API server to function.
+      This policy ensures that mounting of these ServiceAccount tokens is blocked.  
+install: |-
+  ```shell
+  kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/other/restrict-sa-automount-sa-token/restrict-sa-automount-token.yaml
```
Comment (chipzoller) [inline other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml:13]: Has not been updated after change.

File: other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
+name: restrict-sa-automount-sa-token
+version: 1.0.0
+displayName: Restrict Auto-Mount of Service Account Tokens in Service Account Object
```
Comment (chipzoller) [inline other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml:3]: Name has not been updated.

File: other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+createdAt: "2024-01-07T16:15:06.000Z"
```
Comment (chipzoller) [inline other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml:4]: Bump to today.

File: other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      Kubernetes automatically mounts ServiceAccount credentials in each ServiceAccount.
+      The ServiceAccount may be assigned roles allowing Pods to access API resources.
+      Blocking this ability is an extension of the least privilege best practice and should
+      be followed if Pods do not need to speak to the API server to function.
+      This policy ensures that mounting of these ServiceAccount tokens is blocked.  
+  
+  Refer to the documentation for more details on Kyverno annotations: https://artifacthub.io/docs/topics/annotations/kyverno/
+annotations:
+  kyverno/category: "Other"
+  kyverno/kubernetesVersion: "1.23"
```
Comment (chipzoller) [inline other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml:28]: This does not reflect what is in the policy.

**Review by chipzoller — CHANGES_REQUESTED** [review]

File: other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml (hunk 4 above)
Comment (swastik959) [reply other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml:13]: @chipzoller didn't catch what exactly are you trying to say ?

File: other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml (hunk 7 above)
Comment (swastik959) [reply other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml:28]: @chipzoller are you taking about annotations ?

File: other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml (hunk 4 above)
Comment (chipzoller) [reply other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml:13]: Look at the path to the policy file. It's incorrect based on the current state of this PR.

File: other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml (hunk 7 above)
Comment (chipzoller) [reply other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml:28]: Yes, that's where my comment was made, on the `kyverno/kubernetesVersion` annotation. This should be the same value as the corresponding annotation in the policy.

File: other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+readme: |
+      Kubernetes automatically mounts ServiceAccount credentials in each ServiceAccount.
+      The ServiceAccount may be assigned roles allowing Pods to access API resources.
+      Blocking this ability is an extension of the least privilege best practice and should
+      be followed if Pods do not need to speak to the API server to function.
+      This policy ensures that mounting of these ServiceAccount tokens is blocked.  
+  
+  Refer to the documentation for more details on Kyverno annotations: https://artifacthub.io/docs/topics/annotations/kyverno/
+annotations:
+  kyverno/category: "Other"
```
Comment (chipzoller) [inline other/restrict-sa-automount-sa-token/artifacthub-pkg.yaml:27]:
```suggestion
  kyverno/category: "Security"
```

**Review by chipzoller — CHANGES_REQUESTED** [review]

**Review by chipzoller — APPROVED** [review]

**Pull request merged**